File: libs/language/src/lib/utils/limits/PathLimits.ts
import { Limits } from './Limits';
import { BlueNode } from '../../model';
import { NodeToPathLimitsConverter } from './NodeToPathLimitsConverter';

function javaLikeSplit(input: string, delimiter: string | RegExp): string[] {
  const parts = input.split(delimiter);
  const reversedIndex = [...parts].reverse().findIndex((part) => part !== '');
  const cutoff = reversedIndex === -1 ? 0 : parts.length - reversedIndex;
  return parts.slice(0, cutoff);
}

/**
 * Supported features:
 * 1. Exact path matching (e.g., "/a/b/c")
 * 2. Single-level wildcards (e.g., "/a/*\/c")
 * 3. Maximum depth limitation
 */
export class PathLimits extends Limits {
  private readonly allowedPaths: Set<string>;
  private readonly maxDepth: number;
  private readonly currentPath: string[] = [];

  /**
   * Creates path limits with the specified paths and max depth
   * @param allowedPaths - The paths to limit extension to
   * @param maxDepth - The maximum depth of paths to allow
   */
  constructor(allowedPaths: Set<string>, maxDepth: number) {
    super();
    this.allowedPaths = allowedPaths;
    this.maxDepth = maxDepth;
  }

  /**
   * Determines if a path segment should be extended
   * @param pathSegment - The path segment
   * @returns True if the segment should be extended, false otherwise
   */
  public override shouldExtendPathSegment(pathSegment: string): boolean {
    if (this.currentPath.length >= this.maxDepth) {
      return false;
    }

    const potentialPath = this.normalizePath(
      this.getCurrentFullPath() + '/' + pathSegment
    );

    return this.isAllowedPath(potentialPath);
  }

  /**
   * Determines if a path segment should be merged
   * @param pathSegment - The path segment
   * @returns True if the segment should be merged, false otherwise
   */
  public override shouldMergePathSegment(pathSegment: string): boolean {
    return this.shouldExtendPathSegment(pathSegment);
  }

  /**
   * Checks if a path is allowed
   * @param path - The path to check
   * @returns True if the path is allowed, false otherwise
   */
  private isAllowedPath(path: string): boolean {
    for (const allowedPath of this.allowedPaths) {
      if (this.matchesAllowedPath(allowedPath, path)) {
        return true;
      }
    }
    return false;
  }

  /**
   * Checks if a path matches an allowed path pattern
   * @param allowedPath - The allowed path pattern
   * @param path - The path to check
   * @returns True if the path matches the allowed path pattern, false otherwise
   */
  private matchesAllowedPath(allowedPath: string, path: string): boolean {
    const allowedParts = javaLikeSplit(allowedPath, '/');
    const pathParts = javaLikeSplit(path, '/');

    if (pathParts.length > allowedParts.length) {
      return false;
    }

    for (let i = 1; i < pathParts.length; i++) {
      if (allowedParts[i] !== '*' && allowedParts[i] !== pathParts[i]) {
        return false;
      }
    }

    return true;
  }

  /**
   * Enters a path segment
   * @param pathSegment - The path segment
   */
  public override enterPathSegment(pathSegment: string): void {
    this.currentPath.push(pathSegment);
  }

  /**
   * Exits a path segment
   */
  public override exitPathSegment(): void {
    if (this.currentPath.length > 0) {
      this.currentPath.pop();
    }
  }

  /**
   * Gets the current full path
   * @returns The current full path
   */
  private getCurrentFullPath(): string {
    return '/' + this.currentPath.join('/');
  }

  /**
   * Normalizes a path
   * @param path - The path to normalize
   * @returns The normalized path
   */
  private normalizePath(path: string): string {
    return (
      '/' +
      path
        .split('/')
        .filter((s) => s !== '')
        .join('/')
    );
  }

  /**
   * Creates path limits with a maximum depth
   * @param maxDepth - The maximum depth
   * @returns The path limits
   */
  public static withMaxDepth(maxDepth: number): PathLimits {
    const builder = new PathLimitsBuilder().setMaxDepth(maxDepth);
    // Add wildcard patterns for each depth level
    for (let i = 1; i <= maxDepth; i++) {
      const pattern = '/' + Array(i).fill('*').join('/');
      builder.addPath(pattern);
    }
    return builder.build();
  }

  /**
   * Creates path limits with a single path
   * @param path - The path to limit extension to
   * @returns The path limits
   */
  public static withSinglePath(path: string): PathLimits {
    return new PathLimitsBuilder().addPath(path).build();
  }

  /**
   * Creates path limits by analyzing the structure of a node.
   * Leaf paths (no properties and no items) are added as allowed paths.
   */
  public static fromNode(node: BlueNode): PathLimits {
    return NodeToPathLimitsConverter.convert(node);
  }
}

/**
 * Builder for PathLimits
 */
export class PathLimitsBuilder {
  private allowedPaths: Set<string> = new Set<string>();
  private maxDepth: number = Number.MAX_SAFE_INTEGER;

  /**
   * Adds a path to the allowed paths
   * @param path - The path to add
   * @returns The builder
   */
  public addPath(path: string): PathLimitsBuilder {
    this.allowedPaths.add(path);
    return this;
  }

  /**
   * Sets the maximum depth
   * @param maxDepth - The maximum depth
   * @returns The builder
   */
  public setMaxDepth(maxDepth: number): PathLimitsBuilder {
    this.maxDepth = maxDepth;
    return this;
  }

  /**
   * Builds the PathLimits
   * @returns The built PathLimits
   */
  public build(): PathLimits {
    return new PathLimits(this.allowedPaths, this.maxDepth);
  }
}

File: references/java-sdk/src/main/java/blue/language/utils/limits/PathLimits.java
package blue.language.utils.limits;

import blue.language.model.Node;
import blue.language.processor.util.PointerUtils;

import java.util.HashSet;
import java.util.Set;
import java.util.Stack;

/**
 * Supported features:
 * 1. Exact path matching (e.g., "/a/b/c")
 * 2. Single-level wildcards (e.g., "/a/{wildcard}/c")
 * 3. Maximum depth limitation
 */
public class PathLimits implements Limits {
    private final Set<String> allowedPaths;
    private final int maxDepth;
    private final Stack<String> currentPath;

    public PathLimits(Set<String> allowedPaths, int maxDepth) {
        this.allowedPaths = allowedPaths;
        this.maxDepth = maxDepth;
        this.currentPath = new Stack<>();
    }

    @Override
    public boolean shouldExtendPathSegment(String pathSegment, Node node) {
        if (currentPath.size() >= maxDepth) {
            return false;
        }

        String potentialPath = buildPotentialPath(pathSegment);
        return isAllowedPath(potentialPath);
    }

    @Override
    public boolean shouldMergePathSegment(String pathSegment, Node currentNode) {
        return shouldExtendPathSegment(pathSegment, currentNode);
    }

    private boolean isAllowedPath(String path) {
        for (String allowedPath : allowedPaths) {
            if (matchesAllowedPath(allowedPath, path)) {
                return true;
            }
        }
        return false;
    }

    private boolean matchesAllowedPath(String allowedPath, String path) {
        String[] allowedParts = PointerUtils.splitPointerSegments(allowedPath);
        String[] pathParts = PointerUtils.splitPointerSegments(path);

        if (pathParts.length > allowedParts.length) {
            return false;
        }

        for (int i = 0; i < pathParts.length; i++) {
            if (!allowedParts[i].equals("*") && !allowedParts[i].equals(pathParts[i])) {
                return false;
            }
        }

        return true;
    }

    @Override
    public void enterPathSegment(String pathSegment, Node noe) {
        String segment = pathSegment == null ? "" : pathSegment;
        if (segment.startsWith("/")) {
            PointerUtils.validatePointerEscapes(segment);
            currentPath.push(segment.substring(1));
            return;
        }
        if (segment.isEmpty() && currentPath.isEmpty()) {
            return;
        }
        currentPath.push(PointerUtils.escapePointerSegment(segment));
    }

    @Override
    public void exitPathSegment() {
        if (!currentPath.isEmpty()) {
            currentPath.pop();
        }
    }

    private String getCurrentFullPath() {
        return "/" + String.join("/", currentPath);
    }

    private String buildPotentialPath(String pathSegment) {
        String segment = pathSegment == null ? "" : pathSegment;
        if (segment.startsWith("/")) {
            return segment;
        }
        return PointerUtils.appendPointerSegment(getCurrentFullPath(), segment);
    }

    public static class Builder {
        private Set<String> allowedPaths = new HashSet<>();
        private int maxDepth = Integer.MAX_VALUE;

        public Builder addPath(String path) {
            allowedPaths.add(normalizeAndValidateAllowedPath(path));
            return this;
        }

        public Builder setMaxDepth(int maxDepth) {
            if (maxDepth < 0) {
                throw new IllegalArgumentException("Max depth cannot be negative: " + maxDepth);
            }
            this.maxDepth = maxDepth;
            return this;
        }

        public PathLimits build() {
            return new PathLimits(allowedPaths, maxDepth);
        }

        private String normalizeAndValidateAllowedPath(String path) {
            if (path == null) {
                throw new IllegalArgumentException("Allowed path cannot be null");
            }
            String normalized = path.trim();
            if (normalized.isEmpty()) {
                throw new IllegalArgumentException("Allowed path cannot be empty");
            }
            try {
                normalized = PointerUtils.normalizePointer(
                        normalized.startsWith("/") ? normalized : "/" + normalized);
            } catch (IllegalArgumentException ex) {
                throw new IllegalArgumentException("Invalid JSON pointer escape in allowed path: " + normalized);
            }
            return normalized;
        }
    }

    public static PathLimits withMaxDepth(int maxDepth) {
        Builder builder = new PathLimits.Builder().setMaxDepth(maxDepth);
        if (maxDepth <= 0) {
            return builder.addPath("/").build();
        }
        StringBuilder wildcardPath = new StringBuilder();
        for (int i = 0; i < maxDepth; i++) {
            wildcardPath.append("/*");
        }
        return builder.addPath(wildcardPath.toString()).build();
    }

    public static PathLimits withSinglePath(String path) {
        return new PathLimits.Builder().addPath(path).build();
    }

    public static PathLimits fromNode(Node node) {
        return NodeToPathLimitsConverter.convert(node);
    }
}